test/framework/kubernetesversions/template.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetesversions
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cabpkv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
+	kcpv1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/yaml"
+)
+
+type GenerateCIArtifactsInjectedTemplateForDebianInput struct {
+	// ArtifactsDirectory is where conformance suite output will go. Defaults to _artifacts
+	ArtifactsDirectory string
+	// SourceTemplate is an input YAML clusterctl template which is to have
+	// the CI artifact script injection
+	SourceTemplate []byte
+	// PlatformKustomization is an SMP (strategic-merge-style) patch for adding
+	// platform specific kustomizations required for use with CI, such as
+	// referencing a specific image
+	PlatformKustomization []byte
+	// KubeadmConfigTemplateName is the name of the KubeadmConfigTemplate resource
+	// that needs to have the Debian install script injected. Defaults to "${CLUSTER_NAME}-md-0".
+	KubeadmConfigTemplateName string
+	// KubeadmControlPlaneName is the name of the KubeadmControlPlane resource
+	// that needs to have the Debian install script injected. Defaults to "${CLUSTER_NAME}-control-plane".
+	KubeadmControlPlaneName string
```

Comment:
Should we add Machine pool as well?

Reply:
We're only customising the PreKubeadmCommands and Files bits, so we shouldn't need to touch MachinePools do we?

Reply:
if this is the case, why do we need KubeadmConfigTemplateName and KubeadmControlPlaneName?

Reply:
those are the resources that get customised.

Reply:
Please look at https://github.com/kubernetes-sigs/cluster-api/pull/3652/files#diff-646b9b4cde9720e1732b0305f6f337edR174-R190

Reply:
Ah, the KubeadmConfig for a MachinePool. Makes sense. Will PR.

Reply:
https://github.com/kubernetes-sigs/cluster-api/pull/3730